src/Input/Parser.php

```diff
@@ -86,76 +93,213 @@ public function parse(array $argv): self
     }
 
     /**
-     * Parse single arg.
+     * Parse a single arg.
      *
-     * @param string $arg
+     * @param Token $arg The current token
+     * @param Tokenizer $queue The queue of tokens to be consumed
      *
-     * @return mixed
+     * @return void
      */
-    protected function parseArgs(string $arg)
+    protected function parseArgs(Token $arg, Tokenizer $queue) : void
     {
-        if ($this->_lastVariadic) {
-            return $this->set($this->_lastVariadic, $arg, true);
+        // Handle this argument:
+        $argument = array_shift($this->_arguments);
+        
+        //No argument defined, so its an indexed arg:
+        if (is_null($argument)) {
+
+            //Its a single constant value arg:
+            if ($arg->isConstant()) {
+
+                $this->set(null, $arg->value());
+
+            } else {
```

Comment:
the if/else looks a bit oddly. can we refactor it? maybe a method extraction for non-constant case

Reply:
What do you mean? I'm still determining... 

I had another version which was recursive and messy.
I think the naive way of simple, expressive conditions is the simplest and easiest way (and performant) 

Can you elaborate on that?